[wasm] Enable more networking tests

pavelsavara opened the pull request:
- System.Net.Http\tests\FunctionalTests is now included in tests
- I moved System.Net.Http.Functional.Tests and System.Net.WebSockets.Client.Tests to run in chrome instead of V8 via `WasmTestOnBrowser`
- I used `[ActiveIssue]` https://github.com/dotnet/runtime/issues/37669 for missing crypto MD5
- I used `[ActiveIssue]` https://github.com/dotnet/runtime/issues/42852 for loopback support
- I used `[ActiveIssue]` https://github.com/dotnet/runtime/issues/53591 - GET/HEAD method cannot have body
- I used `[ActiveIssue]` https://github.com/dotnet/runtime/issues/53592 - 'TRACE' HTTP method is unsupported.
- I used `[ActiveIssue]` https://github.com/dotnet/runtime/issues/53018 - Failed to fetch
- I used `[ActiveIssue]` https://github.com/dotnet/runtime/issues/53281 for WebSocket problems on Mono/Windows
- I used `[SkipOnPlatform]` when I was confident that we do not plan to support the feature
- Fixed `IsMsQuicSupported` to return bool instead of throwing

akoeplinger commented:
/azp run runtime-libraries-mono

akoeplinger commented:
/azp run runtime-libraries-mono outerloop

akoeplinger commented on src/libraries/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj:
```diff
@@ -6,6 +6,7 @@
     <DefineConstants>$(DefineConstants);SYSNETHTTP_NO_OPENSSL;HTTP3</DefineConstants>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <IncludeRemoteExecutor>true</IncludeRemoteExecutor>
+    <Scenario>WasmTestOnBrowser</Scenario>
```
I think we should have a condition here so it only applies to scenario to browser:
```suggestion
    <Scenario Condition="'$(TargetsBrowser)'=='true'">WasmTestOnBrowser</Scenario>
```

pavelsavara replied:
done

akoeplinger commented on src/libraries/System.Net.Http/tests/FunctionalTests/HttpContentTest.cs:
```diff
@@ -829,6 +836,7 @@ public async Task ReadAsStreamAsync_Buffered_IgnoresCancellationToken(bool readS
         [InlineData(true)]
         [InlineData(false)]
         [ActiveIssue("https://github.com/dotnet/runtime/issues/39187", TestPlatforms.Browser)]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/42852", TestPlatforms.Browser)]
```
#39187 seems to be a different issue? we should replace it with the one you added here

pavelsavara replied:
Both of them possibly apply still.  I didn't try to resolve #39187.
There are at least 7 other places with 39187 on it.
I prefer to keep 39187 open until the whole test passes.

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

mdh1418 commented on src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.MaxConnectionsPerServer.cs:
```diff
@@ -64,6 +64,7 @@ public void Set_ValidValues_Success(int validValue)
         [InlineData(3, 2, true)]
         [InlineData(3, 5, false)]
         [OuterLoop("Uses external servers")]
+        [SkipOnPlatform(TestPlatforms.Browser, "MaxConnectionsPerServer not supported on Browser")]
```
Should `SkipOnPlatform` be added for the rest of the methods that use `MaxConnectionsPerServer`? Or are we only adding it to this method because it is part of the OuterLoop?

We should address the other tests in a followup PR if not in this one.

pavelsavara replied:
The whole class should have it actually and this is a duplicate, thanks for catching it.

mdh1418 commented on src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.RemoteServer.cs:
```diff
@@ -843,6 +869,7 @@ public async Task SendAsync_SendSameRequestMultipleTimesDirectlyOnHandler_Succes
         [OuterLoop("Uses external server")]
         [Theory]
         [MemberData(nameof(Http2Servers))]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/53018", TestPlatforms.Browser)]
```
Does `SendAsync_RequestVersion20_ResponseVersion20` work? Is there something particular about `SendAsync_RequestVersion20_ResponseVersion20IfHttp2Supported` that fails that is not a part of `SendAsync_RequestVersion20_ResponseVersion20`

pavelsavara replied:
some of the HTTP/2 is probably broken, but I want to investigate that under #53018 
The `SendAsync_RequestVersion20_ResponseVersion20` is Windows only test

mdh1418 commented on src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.RemoteServer.cs:
```diff
@@ -94,6 +96,7 @@ public async Task SendAsync_SimpleGet_Success(Configuration.Http.RemoteServer re
 
         [OuterLoop("Uses external server")]
```
Should we change these to `[OuterLoop("Uses external servers")]` as well, like in `src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs`

pavelsavara replied:
unified everywhere

mdh1418 commented on src/libraries/sendtohelixhelp.proj:
```diff
@@ -371,8 +372,8 @@
     <Message Condition="'$(Scenario)' != ''" Importance="High" Text="Done building Helix work items for scenario $(Scenario). Work item count: @(_WorkItem->Count())" />
-    <Message Condition="'$(Scenario)' == '' and ('$(TargetOS)' == 'Android' or '$(TargetOS)' == 'iOS' or '$(TargetOS)' == 'iOSSimulator' or '$(TargetOS)' == 'tvOS' or '$(TargetOS)' == 'tvOSSimulator' or '$(TargetOS)' == 'MacCatalyst')" Importance="High" Text="Done building Helix work items. Work item count: @(XHarnessAppBundleToTest->Count())" />
-    <Message Condition="'$(Scenario)' == '' and '$(TargetOS)' != 'Android' and '$(TargetOS)' != 'iOS' and '$(TargetOS)' != 'iOSSimulator' and '$(TargetOS)' != 'tvOS' and '$(TargetOS)' != 'tvOSSimulator' and '$(TargetOS)' != 'MacCatalyst'" Importance="High" Text="Done building Helix work items. Work item count: @(_WorkItem->Count())" />
+    <Message Condition="'$(Scenario)' == '' and ('$(TargetOS)' == 'Android' or '$(TargetOS)' == 'iOS' or '$(TargetOS)' == 'iOSSimulator' or '$(TargetOS)' == 'tvOS' or '$(TargetOS)' == 'tvOSSimulator' or '$(TargetOS)' == 'MacCatalyst')" Importance="High" Text="Done building Helix work items for scenario $(Scenario). Work item count: @(XHarnessAppBundleToTest->Count())" />
```
wouldn't the text evaluate to ` for scenario .` if the message condition is `'$(Scenario)' == '' and ...`

mdh1418 reviewed (approved):
Thanks, overall it looks good to me!

radical commented on src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs:
```diff
@@ -211,6 +242,12 @@ public async Task PostNonRewindableContentUsingAuth_PreAuthenticate_Success(Conf
         [Theory, MemberData(nameof(RemoteServersMemberData))]
         public async Task PostAsync_EmptyContent_ContentTypeHeaderNotSent(Configuration.Http.RemoteServer remoteServer)
         {
+            if (remoteServer.HttpVersion.Major >= 2 && PlatformDetection.IsBrowser)
```
I think a different issue should be opened for it, if it is not the same failure as others in #53018, with details of the failure.

pavelsavara replied:
I created https://github.com/dotnet/runtime/issues/53591 and https://github.com/dotnet/runtime/issues/53592, for cases which have more specific error message and tagged the relevant tests.

radical commented on src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs:
```diff
@@ -80,11 +95,18 @@ public async Task PostEmptyContentUsingChunkedEncoding_Success(Configuration.Htt
         }
 
         [OuterLoop("Uses external servers")]
+        [SkipOnPlatform(TestPlatforms.Browser, "Not supported on Browser")]
```
What isn't supported here? And same question for the other instances of `"Not supported on Browser"`.

pavelsavara replied:
Actually, we don't know yet. all we get is `Failed to fetch` error, which I would like to address in https://github.com/dotnet/runtime/issues/53018
So I changed this to `ActiveIssue` rather.

radical commented on src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.ServerCertificates.cs:
```diff
@@ -235,7 +235,7 @@ public async Task NoCallback_RevokedCertificate_NoRevocationChecking_Succeeds()
         }
 
-        [OuterLoop("Uses external server")]
+        [OuterLoop("Uses external servers")]
```
Do we really need this change? It makes the diff really hard to read.

pavelsavara replied:
It was suggested by Mitchell we should unify it and I agree. I separated it into different commit on this PR to make it bit easier to review.

radical commented on eng/pipelines/libraries/outerloop-mono.yml:
```diff
@@ -38,6 +38,9 @@ jobs:
         extraStepsParameters:
+          scenarios:
+          - wasmtestonbrowser
+          - normal
```
do we need both browser, and v8 run?

pavelsavara replied:
I think we use v8 for most tests currently because it's lightweight. 
On the other hand, it doesn't have some features we use. This PR is mostly about networking and that's feature of browser, not feature of js engine. NodeJs has separate implementation AFAIK.

radical replied:
Then can we use just browser here? IIUC, `wasmtestonbrowser` will run tests run with `normal`(v8)+the networking ones. Not objecting to it, just making sure that this is the intention.

radical reviewed (changes requested):
I would suggest splitting this PR into two:

1. Enable as many of the networking tests that you are enabling here, *without* the *browseronly* part being needed
2. Add the browseronly bit, and enable the tests that can be enabled because of that

pavelsavara commented:
> I would suggest splitting this PR into two:
> 
> 1. Enable as many of the networking tests that you are enabling here, _without_ the _browseronly_ part being needed
> 2. Add the browseronly bit, and enable the tests that can be enabled because of that

There is not much networking in V8 alone.

radical commented:
> > I would suggest splitting this PR into two:
> > 
> > 1. Enable as many of the networking tests that you are enabling here, _without_ the _browseronly_ part being needed
> > 2. Add the browseronly bit, and enable the tests that can be enabled because of that
> 
> There is not much networking in V8 alone.

Ok, so none of the new tests being enabled now should run with v8? I'm trying to separate this into:

1. enable the 3 test suites, for the regular v8 run, disable any that don't make sense on v8, or don't work
   - is this zero?

2. Add browseronly bit, enable/disable any tests affected by this. Then we can clean this part up too.

pavelsavara commented:
> > > I would suggest splitting this PR into two:
> > > 
> > > 1. Enable as many of the networking tests that you are enabling here, _without_ the _browseronly_ part being needed
> > > 2. Add the browseronly bit, and enable the tests that can be enabled because of that
> > 
> > 
> > There is not much networking in V8 alone.
> 
> Ok, so none of the new tests being enabled now should run with v8? I'm trying to separate this into:
> 
> 1. enable the 3 test suites, for the regular v8 run, disable any that don't make sense on v8, or don't work
>    
>    * is this zero?
> 2. Add browseronly bit, enable/disable any tests affected by this. Then we can clean this part up too.

We have ill-named `IsBrowserDomSupported`. We could start with that.

pavelsavara committed:
- unification of attribute
- fix Quic detection in test
- disable webSocket tests on Mono/Windows
- enable System.Net.Http.Functional.Tests on browser
- disable tests with MaxConnectionsPerServer not supported on Browser
- disable tests with Socket not supported on Browser
- disable tests with AutomaticDecompression not supported on Browser
- disable tests with UseProxy not supported on Browser

pavelsavara committed:
- disable tests with PreAuthenticate not supported on Browser
- disable tests with Credentials is not supported on Browser
- disable tests with Loopback server

pavelsavara committed:
- run outerloop on browser, rather than V8 only
- [ActiveIssue] #37669 for missing crypto MD5

pavelsavara committed:
- conditional tests with PlatformDetection.IsBrowserDomSupportedOrNotBrowser to allow System.Net.Http.Functional.Tests pass on V8

pavelsavara committed:
- System.Net.Http.Functional.Tests to run in chrome instead of V8 via WasmTestOnBrowser

pavelsavara committed:
- disabled tests with https://github.com/dotnet/runtime/issues/53018
- disabled tests with https://github.com/dotnet/runtime/issues/53591
- disabled tests with https://github.com/dotnet/runtime/issues/53592

pavelsavara commented:
- for `System.Net.WebSockets.Client.Tests` it's Condition not met: `WebSocketsSupported` in 100% cases
- for `Common.Tests` you were right, we could keep it in `normal` scenario.
- for `System.Net.WebSockets.Client.Tests` I marked it as `WasmTestOnBrowser` which should run it on browser when available
- I created https://github.com/dotnet/runtime/issues/53591 and https://github.com/dotnet/runtime/issues/53592 with more specific error details for the tests which have specific failure

pavelsavara commented:
>    * is this zero?

For `System.Net.WebSockets.Client.Tests` yes. 
For `System.Net.Http.Functional.Tests` V8 need to skip the same 78 tests as browser, but for different reason. I made that distinction explicit.

pavelsavara commented on src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.Decompression.cs:
```diff
@@ -170,6 +170,7 @@ public static IEnumerable<object[]> DecompressedResponse_MethodNotSpecified_Orig
         }
 
         [OuterLoop("Uses external servers")]
+        [SkipOnPlatform(TestPlatforms.Browser, "AutomaticDecompression not supported on Browser")]
```
@lewing ?

lewing replied:
I don't think we have any control over it at this level.

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

pavelsavara committed:
- ActiveIssue https://github.com/dotnet/runtime/issues/34690 for WebSocket on Windows Mono

pavelsavara committed:
fix pipeline

pavelsavara commented:
/azp run runtime-libraries-mono outerloop

radical commented on src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTestBase.SocketsHttpHandler.cs:
```diff
@@ -16,7 +16,7 @@ public abstract partial class HttpClientHandlerTestBase : FileCleanupTestBase
         protected virtual QuicImplementationProvider UseQuicImplementationProvider => null;
 
-        public static bool IsMsQuicSupported => QuicImplementationProviders.MsQuic.IsSupported;
+        public static bool IsMsQuicSupported => PlatformDetection.IsNotBrowser && QuicImplementationProviders.MsQuic.IsSupported;
```
Should `QuicImplementationProviders.MsQuic.IsSupported` be updated to handle the browser case? That way, if an app tries to use that API, then it would return `false` for them?

pavelsavara replied:
The problem is that QuicImplementationProviders throws PlatformNotSupportedException on AssemblyInfo level.

radical replied:
What do you mean by "on AssemblyInfo level"?

pavelsavara replied:
https://github.com/dotnet/runtime/commit/33d881fd04242635e3d667f93acaad433baa308f#diff-1d4ce147edca0ce81f312b96fce76ac52eb02eb6d04e660817e3b47230edc794R9

radical replied:
IIUC, `QuicImplementationProviders.MsQuic` will throw PNSE. And the supported platforms are at least partially being encoded via the target frameworks in the project file. So, IMHO, instead of explicitly adding `IsNotBrowser` here, I think we should return `QuicImplementationProviders.MsQuic.IsSupported`, and if it throws PNSE, then return `false`.

pavelsavara replied:
Okay

radical commented on src/libraries/Common/tests/System/Net/Http/PostScenarioTest.cs:
```diff
@@ -62,30 +65,52 @@ public async Task PostNoContentUsingContentLengthSemantics_Success(Configuration
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/37669", TestPlatforms.Browser)]
         public async Task PostEmptyContentUsingChunkedEncoding_Success(Configuration.Http.RemoteServer remoteServer)
         {
+            if (remoteServer.HttpVersion.Major >= 2 && PlatformDetection.IsBrowser)
+            {
+                // [ActiveIssue("https://github.com/dotnet/runtime/issues/53018", TestPlatforms.Browser)]
+                return;
+            }
```
Doesn't the `ActiveIssue` disable this for wasm completely? why is this needed?

pavelsavara replied:
you are right, fixed

pavelsavara committed:
feedback

pavelsavara committed:
feedback

radical reviewed (approved):
Thanks for addressing the feedback. Assuming everything passes, LGTM, thanks 👍

pavelsavara commented:
/azp run runtime-libraries-mono outerloop
